### PR TITLE
Enable L1Tracks in L1T DQM (backport)

### PR DESCRIPTION
**Pull request by emacdonald16**
#### PR description:
Backport of #30460

Updating Level-1 Trigger DQM (TTClusters, TTStubs, and TTTracks).
Clusters:
-cluster phi range extended
Stubs:
-stub phi range extended
-5 new distributions added (trigger bend, hardware bend, raw bend, bend offset, and PS module) -> very useful in finding bugs when the TTStub class is updated
Tracks:
-Modified the track input so it will automatically run on the newly merged Hybrid Tracks
-Updated the general track directory to by high-quality ("HQ") tracks and all produced tracks
-Updated the definition of HQ track -> reflects the track quality cuts optimized for Level-1 track MET
-Added nLayersMissed distribution, determined by the L1T group to be very important in separating real from fake tracks
-Added bend chi2 distribution, determined by the L1T group to be very important in separating real from fake tracks
-Added two Chi2 distributions, Chi2 r-phi and Chi2 r-z

Validation package:
-Removed the matched tracks chi2 and chi2dof distributions (they are not very helpful for validation)
-Cleaned up some of the code, removing pieces that are no longer necessary due to the cleaned up TTTrack class (L1Tk_nPar, L1Tk_maxChi2, TP_maxPt, and TP_select_eventid)
**Note** DQM/SiOuterTracker/Track plots are empty, since the running of the L1T tracks are not yet included in CMSSW_11_1_X (at least when I tested it)

#### PR validation:

Ran standalone to make sure output was as expected.
Ran runTheMatrix for 23234.0 

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
Backport of #30460
Requested.

### CODE REVIEW

**Comment by silviodonato** [comment]
please test

**Comment by silviodonato** [comment]
```
20-Jul-2020 11:07:31 CESTSuccessfully opened file file:step3_inDQM.root
%MSG-w FastTimerServiceClient:   FastTimerServiceClient:dqmFastTimerServiceClient@endLumi  20-Jul-2020 11:07:53 CEST Run: 1 Lumi: 1
no lumiVsLS ME is available in HLT/LumiMonitoring
%MSG
%MSG-w FastTimerServiceClient:   FastTimerServiceClient:dqmFastTimerServiceClient@endLumi  20-Jul-2020 11:07:53 CEST Run: 1 Lumi: 1
no puVsLS ME is available in HLT/LumiMonitoring
%MSG
%MSG-e DQMGenericClient:  DQMGenericClient:postProcessorTrack@endRun  20-Jul-2020 11:07:54 CEST End Run: 1
 DQMGenericClient::findAllSubdirectories ==> Missing folder Tracking/TrackBHadron !!!
%MSG
%MSG-e DQMGenericClient:  DQMGenericClient:trackingEffFromHitPattern@endRun  20-Jul-2020 11:08:00 CEST End Run: 1
 DQMGenericClient::findAllSubdirectories ==> Missing folder Tracking/TrackParameters/highPurityTracks/pt_1/HIP_OOT_noINpu !!!
%MSG
%MSG-e DQMGenericClient:  DQMGenericClient:trackingEffFromHitPattern@endRun  20-Jul-2020 11:08:00 CEST End Run: 1
 DQMGenericClient::findAllSubdirectories ==> Missing folder Tracking/TrackParameters/highPurityTracks/pt_1/HIP_noOOT_INpu !!!
%MSG
%MSG-e DQMGenericClient:  DQMGenericClient:trackingEffFromHitPattern@endRun  20-Jul-2020 11:08:00 CEST End Run: 1
 DQMGenericClient::findAllSubdirectories ==> Missing folder Tracking/TrackParameters/highPurityTracks/pt_1/noHIP_noOOT_INpu !!!
%MSG
%MSG-e HcalDigisClient:  HcalDigisClient:hcaldigisClient@endJob  20-Jul-2020 11:08:02 CEST EndJob
No maps histos found...
%MSG
%MSG-e HcalDigisClient:  HcalDigisClient:hcaldigisClient@endJob  20-Jul-2020 11:08:02 CEST EndJob
Error in HcalDigisEndjob! HE
%MSG


A fatal system signal has occurred: segmentation violation
The following is the call stack containing the origin of the signal.

Mon Jul 20 11:08:02 CEST 2020
Thread 2 (Thread 0x2ad6a9f7a700 (LWP 1993)):
#0  0x00002ad689d151d9 in waitpid () from /lib64/libpthread.so.0
#1  0x00002ad6917caa47 in edm::service::cmssw_stacktrace_fork() () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00002ad6917cb50a in edm::service::InitRootHandlers::stacktraceHelperThread() () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00002ad689921d2f in execute_native_thread_routine () at ../../../../../libstdc++-v3/src/c++11/thread.cc:80
#4  0x00002ad689d0dea5 in start_thread () from /lib64/libpthread.so.0
#5  0x00002ad68a0208dd in clone () from /lib64/libc.so.6
Thread 1 (Thread 0x2ad68bdec200 (LWP 1951)):
#0  0x00002ad68a015c3d in poll () from /lib64/libc.so.6
#1  0x00002ad6917caeaf in full_read.constprop () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#2  0x00002ad6917cb5ec in edm::service::InitRootHandlers::stacktraceFromThread() () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#3  0x00002ad6917cc4c9 in sig_dostack_then_abort () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/pluginFWCoreServicesPlugins.so
#4  <signal handler called>
#5  0x00002ad6bb90793f in OuterTrackerMCHarvester::dqmEndJob(dqm::implementation::IBooker&, dqm::implementation::IGetter&) () from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-07-19-2300/lib/slc7_amd64_gcc820/pluginValidationSiOuterTrackerVAuto.so
#6  0x00002ad6bb90bda0 in virtual thunk to DQMEDHarvester::endJob() () from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_1_X_2020-07-19-2300/lib/slc7_amd64_gcc820/pluginValidationSiOuterTrackerVAuto.so
#7  0x00002ad687829cee in edm::Worker::endJob() () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#8  0x00002ad68782bce0 in edm::WorkerManager::endJob(edm::ExceptionCollector&) () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#9  0x00002ad6877e7b95 in edm::Schedule::endJob(edm::ExceptionCollector&) () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#10 0x00002ad6877310e8 in edm::EventProcessor::endJob() () from /cvmfs/cms-ib.cern.ch/nweek-02638/slc7_amd64_gcc820/cms/cmssw/CMSSW_11_1_X_2020-07-19-0000/lib/slc7_amd64_gcc820/libFWCoreFramework.so
#11 0x0000000000412dce in main::{lambda()#1}::operator()() const ()
#12 0x0000000000411362 in main ()

Current Modules:

Module: none (crashed)

A fatal system signal has occurred: segmentation violation

```

**Comment by emacdonald16** [comment]
Huh. Let me investigate. It's not crashing when I run that wf locally...

**Comment by srimanob** [comment]
@emacdonald16 @silviodonato 
It is unclear why we need this PR backport at this point. We don't have a plan to make additional relvals + DQM with 11_1. The plan is to run in production mode for both L1T only small campaign, and L1T+RECO big campaign. Do I miss anything?

However, Looking on PR, there is additional contents to add. I would like to confirm if they are needed for a big ReReco campaign or not.

**Comment by emacdonald16** [comment]
Hey @srimanob, this PR is more for private testing in 11_1, and not for any large (or even really small) campaigns. Since the tracks will be in CMSSW_11_1_0_pre9 and up, it would be nice to have updated tools to quickly check if any bugs accidentally slipped in, rather than for large DQM work.

**Comment by emacdonald16** [comment]
Ok @silviodonato, after multiple attempts, I haven't been able to reproduce the seg fault that caused the 20434.0 wf to fail. So I have reverted the only thing I could think of that MIGHT cause the error back to the way it is now in 11_1. Hopefully that will fix it.

**Comment by silviodonato** [comment]
please test

File: Configuration/EventContent/python/EventContent_cff.py [hunk 1]
```diff
@@ -891,7 +891,8 @@ def SwapKeepAndDrop(l):
         'keep Phase2TrackerDigiedmDetSetVector_mix_*_*',
         'keep *_TTClustersFromPhase2TrackerDigis_*_*',
         'keep *_TTStubsFromPhase2TrackerDigis_*_*',
-        'keep *_TrackerDTC_*_*'
+        'keep *_TrackerDTC_*_*',
+        'keep *_*_Level1TTTracks_*'
```
Comment (silviodonato) [inline Configuration/EventContent/python/EventContent_cff.py:895]: Are you sure you need to add *_*_Level1TTTracks_*? I think the L1 tracks are already included in the FEVTEventContent (they are used for the upcoming re-reco of the HLT TDR samples).

**Comment by silviodonato** [comment]
@emacdonald16 in general this backport might be useful to check that possible future backports does not break L1 tracking. As all backports, this PR should not change the current behaviour of the release apart from adding new features or bug-fixes.

**Comment by srimanob** [comment]
Hi @emacdonald16 
Seem to have an issue with Harvesting step, could you please check? Thanks.

File: Configuration/EventContent/python/EventContent_cff.py (hunk 1 above)
Comment (emacdonald16) [reply Configuration/EventContent/python/EventContent_cff.py:895]: I am not completely sure, no. At the time of the original PR, it was needed, or else no tracks were saved. From digging through the EventContent, it appears the tracks are saved in phase2_muon, but not in phase2_tracker, but both are run in the Phase2 chain. I can definitely remove this and see if things still run well.

**Comment by emacdonald16** [comment]
Checking now. It's not breaking when I run myself, so I'm a bit lost on what the problem is. Is there a way I can do test runs in jenkins?

**Comment by emacdonald16** [comment]
@srimanob and @silviodonato, for the life of me, I cannot reproduce this error. Do either of you have any suggestions?

**Comment by srimanob** [comment]
Hi @emacdonald16 
I can reproduce the error by
```
cmsrel CMSSW_11_1_2_patch1
git cms-merge-topic 30832
scram b -j 8
runTheMatrix.py --what upgrade -l 23234.0 --noCaf --wm init
```
and then run.

You can find root file, and python at
`~srimanob/public/ForDQM/L1T`

To me, it seems you have a problem with
`int testNumEntries3 = resPt1c->GetEntries();`
How do you know that all histograms are filled? 
I check from the ROOT file I provide, and I see resPt6c is empty. That seems to cause an error. I turn
```
if (resPt1c->GetEntries() > 0 && resPt2c->GetEntries() > 0 && resPt3c->GetEntries() > 0 &&
          resPt4c->GetEntries() > 0 && resPt5c->GetEntries() > 0 && resPt6c->GetEntries() > 0) {
```
on, and things seem to work. 

Maybe finding the proper condition to make sure that you fit only non-empty histograms will solve an issue.

HTH.

**Comment by srimanob** [comment]
By the way, I see that you do checking that all histograms are not empty, then fit in master,
https://github.com/cms-sw/cmssw/blob/master/Validation/SiOuterTrackerV/plugins/OuterTrackerMCHarvester.cc#L361-L364

Not clear that by intention, you want the different code in 11_1 or not.

**Comment by emacdonald16** [comment]
OK, I am going to enable it back to the original code, and if we see the same error again, we'll investigate more.

**Comment by silviodonato** [comment]
please test

**Comment by srimanob** [comment]
Something is still odd, will debug more.

**Comment by srimanob** [comment]
Hi @emacdonald16 
I admit that I confuse on code reverting you have done, compared to master.

I think the suggestion is to enable the check on the size of all histograms before start fitting in the same way as master, e.g.
https://github.com/cms-sw/cmssw/blob/master/Validation/SiOuterTrackerV/plugins/OuterTrackerMCHarvester.cc#L403-L404

However, in your reverting, the checking is still the same, only 1 histogram is checked while you try to fit 6.
https://github.com/cms-sw/cmssw/pull/30832/commits/31e0c77aa10fa9b3913540892396b0db903a9132
You remove what your commented out, so basically the check is the same. 

Like, before reverting:
```
// if (resPt1c->GetEntries() > 0 && resPt2c->GetEntries() > 0 && resPt3c->GetEntries() > 0 &&	
//     resPt4c->GetEntries() > 0 && resPt5c->GetEntries() > 0 && resPt6c->GetEntries() > 0) {
```
and after reverting, you just remove these lines. So, no change is done on checking.


Intention? Why the backport need to be different than master in this PR?

**Comment by emacdonald16** [comment]
Hey @srimanob, the backport does not need to be different than the master. The first try for this backport was using those lines, the ones that check all 6 histograms before fitting, and that was giving the same error. I've been trying to think of a more clever way to protect the fitting from failures, but have been unsuccessful.

**Comment by srimanob** [comment]
Hi @emacdonald16 
Could you please reverting the code to be the same as master, introducing all checks as in master? Because when I try to debug twice (8 days ago, and yesterday), those checks are not there and it causes failure. Then we start from that.

To be clear, the debugging I made 8 days ago (before last reverting), the check was in comment, so no check.

Thanks.

**Comment by silviodonato** [comment]
please test

**Comment by srimanob** [comment]
Hi @emacdonald16 

Looking a bit more, it seems possible that the fit function can fail. Then you get null pointer when you try to get fit parameter. For example, files in
> ~srimanob/public/ForDQM/L1T
The histogram resPt6c is filled with underflow only. It can pass GetEntries() because there is an entry filled in the underflow.

You may consider checking first if there is an error from a fit, e.g. using
```
#include "TFitResult.h"
TFitResultPtr r = resPt6c->Fit(fit2, "RS");
Int_t fitStatus = r;
```
then if r=0 means OK, r<0 means there is an error from fitting. If OK, you can use
```
r->Parameter(2);
r->ParError(2);
```

This should give you exactly the same as
```
resPt6c->GetFunction("fit2")->GetParameter(2)
resPt6c->GetFunction("fit2")->GetParError(2)
```

HTH.

**Comment by emacdonald16** [comment]
Hey @srimanob, thank you so much for your help. Let's see if this works!

**Comment by jfernan2** [comment]
please test

**Comment by silviodonato** [comment]
I'm converting this PR as a draft just to remove this from the ORP spreadsheet. Of course feel free to open it back if needed.

**Comment by silviodonato** [comment]
if you want to open it back, please let me know

**Pull request closed without merging**